Phase-field model for fracture using MPI and AMR

WasimNiyazMunshi opened the pull request:
This pull request adds a phase-field solver for fracture to the code gallery. The fracture model uses the MPI and Adaptive Mesh Refinement (AMR) capabilities of deal.II to solve large scale problems in 3D.

bangerth commented:
@WasimNiyazMunshi is a visiting student in my lab, and I've contributed to this program. I would appreciate if someone else would handle this PR!

bangerth commented on Phase_field_fracture_model_in_3D/phase_field.cc:
```diff
@@ -0,0 +1,1521 @@
+      LA::MPI::SparseMatrix system_matrix_elastic;
+      LA::MPI::Vector locally_relevant_solution_elastic;
+      LA::MPI::Vector completely_distributed_solution_elastic_old;
+      LA::MPI::Vector completely_distributed_solution_elastic;
```
For this and the corresponding `_damage` vector, the only place where you actually use it is in the `solve()` function. I think you should move the vector declaration there. That's because otherwise you store the same information twice, once in `locally_relevant_solution_elastic` and once in `completely_distributed_solution_elastic`, and it's just too easy to make mistakes where the two objects go out of sync. The general rule ought to be to only ever store information in one place.

WasimNiyazMunshi replied:
@bangerth For which of the vectors are you suggesting to move the declaration in the solve()? All the 3 vectors declared above are used in more than one function in the code, so I am not sure if I can move the declaration to solve().

bangerth replied:
I looked through this again, and you're right. I misread your code.

bangerth commented on Phase_field_fracture_model_in_3D/phase_field.cc:
```diff
@@ -0,0 +1,1521 @@
+      AffineConstraints<double> constraints_elastic;
+      LA::MPI::SparseMatrix system_matrix_elastic;
+      LA::MPI::Vector locally_relevant_solution_elastic;
+      LA::MPI::Vector completely_distributed_solution_elastic_old;
```
This variable is only used in the solution transfer process. Move the variable declaration there.

WasimNiyazMunshi replied:
@bangerth Are you referring to 'completely_distributed_solution_elastic_old'. I am using this in 3 functions namely, refine_grid(), check_convergence and run().

marcfehling commented on Phase_field_fracture_model_in_3D/phase_field.cc:
```diff
@@ -0,0 +1,1521 @@
+        constraints_damage);
+
+    // Create initial crack, if any
+    /*for (const auto &cell : dof_handler_damage.active_cell_iterators())
+     if (cell->is_locally_owned())
+
+     {
+     //for (const auto &face : cell->face_iterators())
+
+     for (const auto vertex_number : cell->vertex_indices())
+     {
+     const auto vert = cell->vertex(vertex_number);
+     const Point<3>& node = vert;
+     if (((std::fabs((z_max*(node(0)-0.5*x_max + A)) - 2*A*( node(2))) <10*l)
+     && node(1)>=0.9*y_max)
+     || ((std::fabs((node(0)-0.5*x_max + A)*z_max+2*A*(node(2)-z_max))<10*l)
+     && node(1)<=0.1*y_max))
+     if ((vert(0) - 0.5*(x_min+x_max) < 1e-12) &&
+     (std::fabs(vert(1) - 0.5*(y_min + y_max)) <= bandwidth))  // nodes on initial damage plane
+
+
+     {
+     const unsigned int dof = cell->vertex_dof_index(vertex_number, 0);
+     constraints_damage.add_line(dof);
+     constraints_damage.set_inhomogeneity(dof,1);
+     }
+
+     }
+
+     }*/
```
same here

WasimNiyazMunshi replied:
The problem considered here does not have any initial crack (damage). But often times, the problems one solves involves a pre-existing crack. This part of the code sets up the necessary constraints on the damage field for the initial crack.

marcfehling commented on Phase_field_fracture_model_in_3D/phase_field.cc:
```diff
@@ -0,0 +1,1521 @@
+    TimerOutput::Scope ts (computing_timer, "solve_linear_system_damage");
+
+    SolverControl solver_control (10000,
+        1e-12/** system_rhs_damage.l2_norm()*/);
```
```suggestion
        1e-12 * system_rhs_damage.l2_norm());
```
same here

WasimNiyazMunshi replied:
Fixed now.

marcfehling commented on Phase_field_fracture_model_in_3D/phase_field.cc:
```diff
@@ -0,0 +1,1521 @@
+  PhaseField::load_disp_calculation (const unsigned int load_step)
+
+  {
+    // TODO: You mark these vectors as 'static' because you want them to carry state from one call to this function to the next. That is exactly what member variables are there for -- so make these vectors member variables.
+    static Vector<double> load_values_x (num_load_steps + 1);
+    static Vector<double> load_values_y (num_load_steps + 1);
+    static Vector<double> displacement_values (num_load_steps + 1);
```
I support this statement :+1:

WasimNiyazMunshi replied:
@bangerth  @marcfehling I understand that the ToDo statement suggests making these vectors member variables of the **PhaseField class.** However, since these vectors are only used in this specific function, I thought it would be more appropriate to declare them as local variables instead. Suggestions?

bangerth replied:
`static` local variables are like global variables, with all the problems that entails. In particular, it means that you cannot have two objects of your class at the same time since they would reference the same global variable. I think it is better to make these variables member variables of your class.

WasimNiyazMunshi replied:
Thanks, Wolfgang. I understood where the problem lies. I have declared these as member variables of the class now.

bangerth replied:
As a matter of fact, while you're at it, you might want to declare these as `std::vector` instead of `Vector<double>`. These vectors represent *collections* of values, not vectors in a high-dimensional space that you want to add together, or take dot products, etc. That would also enable you to use a `std::vector<Tensor<1,dim>> load_values` if you wanted to.

marcfehling reviewed (approved):
Nice work. The code looks fine! I did not check the formulas in the code, as they overstep my expertise. According to your manuscript, you validated your code against literature data, which suffices for me.

The following two points are more suggestions than requirements for being merged. Please apply according to your own assessment :)
- If you'd like, you can run a code indentation tool like `clang-format` over your source file, which further improves readability.
- If you are feeling fancy: Consider collecting all parameters in a parameter class or multiple classes, and use the `ParameterHandler`.

blaisb commented on Phase_field_fracture_model_in_3D/phase_field.cc:
```diff
@@ -0,0 +1,1503 @@
+
+    for (const auto &cell : dof_handler_elastic.active_cell_iterators ())
+      if (cell->is_locally_owned ())
+
```
```suggestion
```

blaisb commented on Phase_field_fracture_model_in_3D/phase_field.cc:
```diff
@@ -0,0 +1,1503 @@
+              const double d = damage_values[q_point];
+
+              for (const unsigned int i : fe_values_elastic.dof_indices ())
+
```
```suggestion
```

blaisb commented on Phase_field_fracture_model_in_3D/phase_field.cc:
```diff
@@ -0,0 +1,1503 @@
+  void
+  PhaseField::assemble_system_damage ()
+  {
+
```
```suggestion
```

blaisb reviewed (approved):
Just a few comments there and there. Very cool work!

WasimNiyazMunshi commented on Phase_field_fracture_model_in_3D/phase_field.cc:
```diff
@@ -0,0 +1,1521 @@
+            }
+
+          // traction contribution to rhs
+          /* for (const auto &face : cell->face_iterators())
+           {
+           if (face->at_boundary() &&
+           face->boundary_id() == 10)
+
+           {
+           fe_face_values_elastic.reinit(cell, face);
+           for (const auto i : fe_face_values_elastic.dof_indices())
+
+           {const unsigned int component_i =
+           fe_elastic.system_to_component_index(i).first;
+           for (const auto face_q_point :
+           fe_face_values_elastic.quadrature_point_indices())
+           {
+
+           cell_rhs_elastic(i) +=
+           fe_face_values_elastic.shape_value( i, face_q_point)
+           *(traction_values_elastic[face_q_point][component_i])
+           *fe_face_values_elastic.JxW(face_q_point);
+
+
+
+           }
+           }
+           }
+           }*/
```
This part of the code is used to apply traction boundary conditions. We do not any have traction boundary conditions in the example considered here but I think it would be a good idea to keep this part there incase one wants to solve problems with traction boundary conditions.

WasimNiyazMunshi commented on Phase_field_fracture_model_in_3D/phase_field.cc:
```diff
@@ -0,0 +1,1503 @@
+              else if (d == 2)
+                {
+                  centers_list[i][d] = static_cast<double> (5.0
+                      + ((rand ()) / RAND_MAX) * (8.0 - 5.0));
```
@bangerth I tried to adress this but x_max etc are not recognized in the given scope. So, I was thinking of declaring x_max etc  as static members. Is that a good idea or is there a better way to do this?

bangerth replied:
Perhaps do something like this at the top:
```
namespace Domain
{
  const double x_max = ...;
  // similar for the other variables that describe the domain
}
```
Then you can reference them from anywhere in the program via `Domain::x_max`.

WasimNiyazMunshi replied:
Thanks, Wolfgang. I have fixed it as suggested.

bangerth commented on Phase_field_fracture_model_in_3D/phase_field.cc:
```diff
@@ -0,0 +1,1464 @@
+#include <deal.II/base/quadrature_lib.h>
```
Can you copy the usual license header into this file as well?

bangerth commented on Phase_field_fracture_model_in_3D/phase_field.cc:
```diff
@@ -0,0 +1,1473 @@
+    error_elastic_solution = error_elastic_solution_numerator
+        / error_elastic_solution_denominator;
+
+    completely_distributed_solution_elastic_old =
+        locally_relevant_solution_elastic;
+    completely_distributed_solution_damage_old =
+        locally_relevant_solution_damage;
```
Move these lines to run().

bangerth commented on Phase_field_fracture_model_in_3D/phase_field.cc:
```diff
@@ -0,0 +1,1473 @@
+                    locally_relevant_solution_elastic;
+              }
+            else
+              stoppingCriterion = check_convergence ();
```
Move lines here.

bangerth replied:
Then transform
```
  if (iteration == 0)
    XXXX
  else
    {
       check_convergence();
       XXXX
    }
```
to
```
  if (iteration > 0)
    check_convergence();

  XXXX;

WasimNiyazMunshi replied:
Fixed now.

WasimNiyazMunshi commented:
@blaisb, @Rombur  and @marcfehling. 
Thank you all for taking the time to review my code. I have worked on addressing most of your comments and suggestions. The only pending items are the two blocks of commented-out code that @marcfehling suggested removing. May I kindly request you to review the code once more, so we can proceed with adding it to the code gallery soon? :-) 
Best regards
Wasim

bangerth commented:
As for the commented-out code blocks: @WasimNiyazMunshi suggests to leave them in there because many benchmarks are described with traction boundary conditions rather than displacements, and the commented out code is what you will need if you want to use the code for a traction-based modification. That seems reasonable to me.

WasimNiyazMunshi commented:
Yes. Also, the other block of commented out code is used to describe any
pre-existing cracks. This is very common in most of the benchmark problems
in fracture, so I suggest to keep that as well.


On Sun, 15 Dec, 2024, 16:44 Wolfgang Bangerth, ***@***.***>
wrote:

> As for the commented-out code blocks: @WasimNiyazMunshi
> <https://github.com/WasimNiyazMunshi> suggests to leave them in there
> because many benchmarks are described with traction boundary conditions
> rather than displacements, and the commented out code is what you will need
> if you want to use the code for a traction-based modification. That seems
> reasonable to me.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dealii/code-gallery/pull/203#issuecomment-2544191454>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BLC5F7EQ5LVU43GAL5AR3GT2FYH55AVCNFSM6AAAAABQCLWRZWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKNBUGE4TCNBVGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

marcfehling commented:
> The only pending items are the two blocks of commented-out code that @marcfehling suggested removing.

I'm fine with leaving the comments inside. The comments you placed in front of the code blocks are helpful for context.

marcfehling commented:
You now have a name conflict: The name `PF` was already used in petsc. Your namespace `PF` only contains the `PhaseField` class -- I would suggest to remove your namespace. @WasimNiyazMunshi @bangerth -- What do you think?
```
[ 50%] Building CXX object CMakeFiles/phase_field.dir/phase_field.cc.o
/__w/code-gallery/code-gallery/Phase_field_fracture_model_in_3D/phase_field.cc:60:11: error: ‘namespace PF { }’ redeclared as different kind of entity
   60 | namespace PF
      |           ^~
In file included from /usr/include/petsc/petscdmda.h:6,
                 from /usr/include/petsc/petsc.h:14,
                 from /usr/local/include/deal.II/fe/fe_values.h:53,
                 from /__w/code-gallery/code-gallery/Phase_field_fracture_model_in_3D/phase_field.cc:40:
/usr/include/petsc/petscpf.h:36:23: note: previous declaration ‘typedef struct _p_PF* PF’
   36 | typedef struct _p_PF* PF;
      |                       ^~
/__w/code-gallery/code-gallery/Phase_field_fracture_model_in_3D/phase_field.cc: In function ‘int main(int, char**)’:
/__w/code-gallery/code-gallery/Phase_field_fracture_model_in_3D/phase_field.cc:1437:23: error: ‘PF’ is not a namespace-name
 1437 |       using namespace PF;
      |                       ^~
make[2]: *** [CMakeFiles/phase_field.dir/build.make:76: CMakeFiles/phase_field.dir/phase_field.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:90: CMakeFiles/phase_field.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```

bangerth commented:
We use namespaces in all tutorial programs. Perhaps keep it the same here? Maybe just rename it to `namespace FracturePropagation` or some such?

WasimNiyazMunshi commented:
I agree with you, Wolfgang. I'll make the necessary changes and rename the
namespace.

Thanks
Wasim

On Wed, 8 Jan, 2025, 09:05 Wolfgang Bangerth, ***@***.***>
wrote:

> We use namespaces in all tutorial programs. Perhaps keep it the same here?
> Maybe just rename it to namespace FracturePropagation or some such?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dealii/code-gallery/pull/203#issuecomment-2578048491>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BLC5F7DDBSU2G6JK7HMVKRD2JVEDJAVCNFSM6AAAAABQCLWRZWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKNZYGA2DQNBZGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

WasimNiyazMunshi commented:
Dear All, It seems that the code has now been merged into the master. Thank
you, everyone, for reviewing the code and providing your valuable feedback.

Best regards
Wasim


On Thu, 9 Jan, 2025, 07:23 Marc Fehling, ***@***.***> wrote:

> Merged #203 <https://github.com/dealii/code-gallery/pull/203> into master.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dealii/code-gallery/pull/203#event-15868745227>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BLC5F7D77AACSH3WGIXCWVT2J2A55AVCNFSM6AAAAABQCLWRZWVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJVHA3DQNZUGUZDENY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
